libslide/src/partial_evaluator/types.rs

```diff
@@ -0,0 +1,63 @@
+use std::ops;
+use crate::parser::types::Expr;
+
+pub struct PEResult {
+    pub unevaluated_expr: Option<Box<Expr>>,
+    pub evaluated: f64
+}
```

Comment:
I think this would be better expressed as an enum like `Result` is:

```rust
pub enum PEResult {
  Evaluated(f64),
  Unevaluated(Box<Expr>),
}
```

this way when we need to add two `PEResult`s, we just do something like

```rust
impl ops::Add for PEResult {
  type Output = PEResult;

  fn add(self, rhs: PEResult) -> Self::Output {
    fold_binary(self, rhs, Operator::Plus)
  }
}

fn operator_table(op: Operator) -> Box<dyn Fn(f64, f64) -> f64 + 'static> {
  match op {
    Plus => Box::new(|x, y| x + y),
  }
}

fn fold_binary(lhs: PEResult, rhs: PEResult, op: Operator) -> PEResult {
  match lhs, rhs {
    Evaluated(x), Evaluated(y) => Evaluated(operator_table(op)(x, y)),
    Evaluated(x), Unevaluated(y) => Unevaluated(BinaryOperator{
      op,
      lhs: Number(x),
      rhs: y
    }),
    ...etc
  }
}
```

I hope this makes sense. One thing you may be wondering is why we would want to do dynamic allocation with a box for a function; I think this will not incur a runtime penalty because the `'static` lifetime is allocated for the lifetime of the program IIRC.

Reply:
why is this better?